src/Routing.purs

```diff
@@ -1,67 +1,92 @@
-
-
--- | Stream of hash changed, callback called when new hash can be matched
--- | First argument of callback is `Just a` when old hash can be matched
--- | and `Nothing` when it can't.
-matches :: forall e a. Match a -> (Maybe a -> a -> Eff e Unit) -> Eff e Unit
-matches = matches' decodeURIComponent
-
-matches' :: forall e a. (String -> String) ->
-            Match a -> (Maybe a -> a -> Eff e Unit) -> Eff e Unit
-matches' decoder routing cb = hashes $ \old new ->
-  let mr = matchWith decoder routing
-      fst = either (const Nothing) Just $ mr old
-  in either (const $ pure unit) (cb fst) $ mr new
+-- | Folds effectfully over hash changes given a callback and an initial hash.
+-- | The provided String is the hash portion of the `Location` with the '#'
+-- | prefix stripped. Returns an effect which will remove the listener.
+foldHashes
+  :: forall eff a
+   . (a -> String -> Eff (RoutingEffects eff) a)
+  -> (String -> Eff (RoutingEffects eff) a)
+  -> Eff (RoutingEffects eff) (Eff (RoutingEffects eff) Unit)
+foldHashes cb init = do
+  ref <- newRef =<< init =<< getHash
+  win <- windowToEventTarget <$> window
+  let listener = eventListener \_ -> writeRef ref =<< join (cb <$> readRef ref <*> getHash)
+  addEventListener hashchange listener false win
+  pure $ removeEventListener hashchange listener false win
 
-matchesAff' :: forall e a. (String -> String) ->
-               Match a -> Aff e (Tuple (Maybe a) a)
-matchesAff' decoder routing =
-  makeAff \k -> nonCanceler <$
-    matches' decoder routing \old new ->
-      k $ Right $ Tuple old new
+-- | Runs the callback on every hash change providing the previous hash and the
+-- | latest hash. The provided String is the hash portion of the `Location` with
+-- | the '#' prefix stripped. Returns an effect which will remove the listener.
+hashes
+  :: forall eff
+   . (Maybe String -> String -> Eff (RoutingEffects eff) Unit)
+  -> Eff (RoutingEffects eff) (Eff (RoutingEffects eff) Unit)
+hashes = matchesWith Just
 
-matchesAff :: forall e a. Match a -> Aff e (Tuple (Maybe a) a)
-matchesAff = matchesAff' decodeURIComponent
+-- | Runs the callback on every hash change using a given `Match` parser to
+-- | extract a route from the hash. If a hash fails to parse, it is ignored.
+-- | To avoid dropping hashes, provide a fallback alternative in your parser.
+-- | Returns an effect which will remove the listener.
+matches
+  :: forall eff a
+   . Match a
+  -> (Maybe a -> a -> Eff (RoutingEffects eff) Unit)
+  -> Eff (RoutingEffects eff) (Eff (RoutingEffects eff) Unit)
+matches matcher = matchesWith (matchWith decodeURIComponent matcher)
 
+-- | Runs the callback on every hash change using a given custom parser to
+-- | extract a route from the hash. If a hash fails to parse, it is ignored.
+-- | To avoid dropping hashes, provide a fallback alternative in your parser.
+-- | Returns an effect which will remove the listener.
+matchesWith
+  :: forall eff f a
+   . Foldable f
```

Comment:
I've generalized this to any Foldable, where we just take the first value. Do you think there's something better?

Reply:
I like `Foldable`! (as `List` based parsers could be used too). Using something like https://github.com/purescript/purescript-foldable-traversable/issues/82 instead of `find (const true)` would be nice but we don't have it yet so works fine too.

Reply:
I have opened PR https://github.com/purescript/purescript-foldable-traversable/pull/83 to add `indexl` and `indexr` which will be nice to use here instead of `find`

Reply:
Just a couple of thoughts about this

+ The `f` should have notion of `fail` and `extract`, right? So, I'd say this is `Unfoldable`, but for some reason there is no `Unfoldable` for `Maybe` :confused: 
+ OTOH
```purescript
matchesWith
  :: forall f a 
   . Alternative f 
  => (String -> f a) 
  -> (f a -> a -> Eff _ Unit)
  -> Eff _ (Eff _ Unit)
matchesWith parser cb = foldHashes go $ go empty
  where 
  go = ...
```

Reply:
I don't think `Unfoldable` is quite right. Or I think the accumulator is what would be `Unfoldable`, and not the resulting `f`. `Alternative` is closer, but you need to be able to get the `a` out of it. I guess it could be `Traversable` as well.

Reply:
Heh, we have either `fail` or value, that's precisely `α + 1` --> `Maybe`. What's wrong with `Maybe` btw?

Reply:
Nothing really. It's just that most parsers are going to emit `Either`s but all I care about is success or failure. It just means throwing `hush` on everything.

Reply:
I'm for `Maybe` here. I can't find any sensible constraint semantics for `parser` :)

Reply:
And if you need to deal with empty or extract a value, `fold` is also appropriate, but I don't want to require a `Monoid`.

Reply:
Oh! `Either`, right :)

Reply:
@cryogenian someone could built Lazy.List based parser too, which will just work.

Reply:
If we agree on using Foldable, and if merging it is not urgent, can we wait while https://github.com/purescript/purescript-foldable-traversable/pull/83 is merged?